chain/events/filter/index.go

```diff
@@ -56,27 +61,203 @@ var ddls = []string{
+	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
 )
 
 type EventIndex struct {
 	db *sql.DB
+
+	stmtEventExists          *sql.Stmt
+	stmtInsertEvent          *sql.Stmt
+	stmtInsertEntry          *sql.Stmt
+	stmtRevertEventsInTipset *sql.Stmt
+	stmtRestoreEvent         *sql.Stmt
+}
+
+func (ei *EventIndex) initStatements() (err error) {
+	ei.stmtEventExists, err = ei.db.Prepare(eventExists)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtEventExists: %w", err)
+	}
+
+	ei.stmtInsertEvent, err = ei.db.Prepare(insertEvent)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtInsertEvent: %w", err)
+	}
+
+	ei.stmtInsertEntry, err = ei.db.Prepare(insertEntry)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtInsertEntry: %w", err)
+	}
+
+	ei.stmtRevertEventsInTipset, err = ei.db.Prepare(revertEventsInTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtRevertEventsInTipset: %w", err)
+	}
+
+	ei.stmtRestoreEvent, err = ei.db.Prepare(restoreEvent)
+	if err != nil {
+		return xerrors.Errorf("prepare stmtRestoreEvent: %w", err)
+	}
+
+	return nil
+}
+
+func (ei *EventIndex) migrateToVersion2(ctx context.Context, chainStore *store.ChainStore) error {
+	if chainStore == nil {
+		return errors.New("chainStore is nil")
+	}
+
+	now := time.Now()
+
+	var err error
+	var tx *sql.Tx
+	tx, err = ei.db.Begin()
+	if err != nil {
+		return xerrors.Errorf("begin transaction: %w", err)
+	}
+
+	// in case of error, rollback the transaction
+	defer func() {
+		if err != nil {
+			log.Errorf("rolling back transaction due to: %w", err)
+
+			err2 := tx.Rollback()
+			if err2 != nil {
+				log.Errorf("error in rollback: %s", err2)
+			}
+		}
+	}()
+
+	stmtDeleteOffChainEvent, err := tx.Prepare("DELETE FROM event WHERE tipset_key_cid!=? and height=?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtDeleteOffChainEvent: %w", err)
+	}
+
+	stmtSelectEvent, err := tx.Prepare("SELECT id FROM event WHERE tipset_key_cid=? ORDER BY message_index ASC, event_index ASC, id DESC LIMIT 1")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtSelectEvent: %w", err)
+	}
+
+	stmtDeleteEvent, err := tx.Prepare("DELETE FROM event WHERE tipset_key_cid=? AND id<?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtDeleteEvent: %w", err)
+	}
+
+	// get the lowest height tipset
+	var minHeight sql.NullInt64
+	err = ei.db.QueryRow("SELECT MIN(height) FROM event").Scan(&minHeight)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil
+		}
+
+		return xerrors.Errorf("query min height: %w", err)
+	}
+
+	currTs := chainStore.GetHeaviestTipSet()
+	log.Infof("Migrating events from head to %d", minHeight.Int64)
+
+	for {
+		if int64(currTs.Height()) < minHeight.Int64 {
+			break
+		}
+
+		tsKey := currTs.Parents()
+		var ts *types.TipSet
+		ts, err = chainStore.GetTipSetFromKey(ctx, tsKey)
```

Comment:
```suggestion
		currTs, err = chainStore.GetTipSetFromKey(ctx, tsKey)
```

This lets us avoid re-assigning currTs below. Unless we want to process the first tipset (we aren't currently doing that anyways).